services/offline/src/lib/cacheable-section-state.js

```diff
@@ -0,0 +1,183 @@
+import PropTypes from 'prop-types'
+import React from 'react'
+import {
+    createStore,
+    useGlobalState,
+    useGlobalStateMutation,
+    GlobalStateProvider,
+} from './global-state-service'
+import { useOfflineInterface } from './offline-interface'
+
+// Functions in here use the global state service to manage cacheable section
+// state in a performant way
+
+/**
+ * Helper that transforms an array of cached section objects from the IndexedDB
+ * into an object of values keyed by section ID
+ *
+ * @param {Array} list - An array of section objects
+ * @returns {Object} An object of sections, keyed by ID
+ */
+function transformSections(sectionsArray) {
+    return sectionsArray.reduce(
+        (result, { sectionId, lastUpdated }) => ({
+            ...result,
+            [sectionId]: lastUpdated,
+        }),
+        {}
+    )
+}
+
+/**
+ * Create a store for Cacheable Section state.
+ * Expected to be used in app adapter
+ */
+export function createCacheableSectionStore() {
+    const initialState = { recordingStates: {}, cachedSections: {} }
+    return createStore(initialState)
+}
+
+/**
+ * Helper hook that returns a value that will persist between renders but makes
+ * sure to only set its initial state once.
+ * See https://gist.github.com/amcgee/42bb2fa6d5f79e607f00e6dccc733482
+ */
+function useConst(factory) {
+    const ref = React.useRef(null)
+    if (ref.current === null) ref.current = factory()
+    return ref.current
+}
+
+/**
+ * Provides context for a global state context which will track cached
+ * sections' status and cacheable sections' recording states, which will
+ * determine how that component will render. The provider will be a part of
+ * the OfflineProvider.
+ */
+export function CacheableSectionProvider({ children }) {
+    const offlineInterface = useOfflineInterface()
+    const store = useConst(createCacheableSectionStore)
+
+    // On load, get sections and add to store
+    React.useEffect(() => {
+        offlineInterface.getCachedSections().then(sections => {
+            const sectionsById = transformSections(sections)
+            store.mutate(state => ({
+                ...state,
+                cachedSections: sectionsById,
+            }))
```

Comment:
(*super minor change*)

`transformSections` is a bit vague - `groupSectionsById` or `getSectionsById` will not only make the intent clearer but also allow you to reduce this part (and where it's repeated below) by one line:

```suggestion
            store.mutate(state => ({
                ...state,
                cachedSections: groupSectionsById(sections),
            }))
```

Reply:
Sounds good ~ I saved us another line the second time the function is used too 😉